servers/zts/src/main/java/com/yahoo/athenz/zts/store/DataStore.java

```diff
@@ -723,16 +723,32 @@ void processTrustMembership(DataCache data, String identity, String rolePrefix,
     public DataCache getDataCache(String domainName) {
         return getCacheStore().getIfPresent(domainName);
     }
-    
+
     // API
     public void getAccessibleRoles(DataCache data, String domainName, String identity,
             String roleName, Set<String> accessibleRoles, boolean keepFullName) {
+        getAccessibleRoles(data, domainName, identity, roleName, accessibleRoles, keepFullName, null);
+    }
+    
+    // API additional check for application ID
+    public void getAccessibleRoles(DataCache data, String domainName, String identity,
+            String roleName, Set<String> accessibleRoles, boolean keepFullName, String applicationId) {
 
         /* if the domain hasn't been processed then we don't have anything to do */
         
         if (data == null) {
             return;
         }
+        
+        /* check if application id matches the principal's application id */
```

Comment:
Any reason why we can't do the check one layer above in the ZMSImpl?

Reply:
We can do that in ZMSImpl. I think the reason for this to be in getAccessibleRoles is that I noticed that this api is called through out other apis in ZMSImpl which might need this check as well. If not, I can directly add it to getRoleTokens api.

Reply:
Actually in all those other cases we don't want to add the restriction. It must only be checked when the actual getRoleToken api is called since that's the only time that we sign and return a role token.

Reply:
done.